### DISCUSSION TITLE
How to preserve format when no rule is offended

### DISCUSSION THREAD

**Discussion by ftucky** [scalameta/scalafmt#3727] [Q&A] [answered]
Consider the following code snippet:

```
final def dispatchVisit(cell: Domain): Visit =
    cell match
      case c: OpsPrimary[t]#OpsPrimaryMarker => dispatchOpsPrimary(c)
      case c: Ops       [t]#OpsMarker        => dispatchOps       (c)
```

It *does*  fulfill all formatting rules, including the alignment on the `=>` within a `case`.
The author has been kind enough to enhance code readability by providing *more* alignments than strictly required.

However, `scalafmt` insists in rewriting:
```
  final def dispatchVisit(cell: Domain): Visit =
    cell match
      case c: OpsPrimary[t]#OpsPrimaryMarker => dispatchOpsPrimary(c)
      case c: Ops[t]#OpsMarker               => dispatchOps(c)
```
losing the valuable "extra" formatting.

`scalafmt` also (wrongly) claims that the original code is not properly formatted:
```
[warn] scalafmt: xxx.scala isn't formatted properly!
```

(a) **Is there any `scalafmt` configuration which would preserve the original format ?**

Using format `\\ format: off` is not a valid response. This is a perfectly nominal case, which does not require a particular waiver.
The point is neither to configure `scalafmt` to find the correct alignment by itself. Often the correct formatting reflects code author's *intention*, which is not accessible to `scalafmt`.  

(b) A `scalafmt` configuration mostly define rules:
* lines shall not be longer than than xxx characters
* breaks shall be encoded as xxx
* indententation shall be x spaces / y tabs 
* paren shall be aligned
* ...

** Should `scalafmt` report that a file is not `formatted properly` only when one of those rules is violated ? ** 

(c) **If "no", would it make sense ?**

For some configurations like `newlines.source` scalafmt proposes to `keep` the original choices.
Should it be available for `align` or `spaces` ?

**Comment by kitbellew** [comment, marked as answer]
@ftucky there's an implied additional "rule" (or property of scalafmt design), that for everything else there's a predefined fallback (no space is some cases, one space in others).

scalafmt is not a linter which checks compliance with some rules, it's a tool which will produce its own formatting of code subject to a few configurable partial constraints.

if scalafmt is asked to verify formatting, it checks whether the code was previously formatted by scalafmt using the same configuration, and not whether it complies with a subset of intentions expressed via a few explicit parameters in that configuration.

it's impossible to do that you described.

**Reply by ftucky** [reply]
OK, point taken. Thank you.

Note that `scalafmtCheckAll` very much *looks* like a format linter !
